module/pkgs-by-name-for-flake-parts: init

drupol opened the pull request:
New module: https://github.com/drupol/pkgs-by-name-for-flake-parts

roberth commented on flake.nix:
```diff
@@ -374,15 +388,42 @@
-          baseUrl = "https://github.com/cachix/git-hooks.nix/blob/master";
+        pkgs-by-name-for-flake-parts = {
+          title = "pkgs-by-name-for-flake-parts";
+          baseUrl = "https://github.com/drupol/pkgs-by-name-for-flake-parts";
```
```suggestion
          baseUrl = "https://github.com/drupol/pkgs-by-name-for-flake-parts/blob/master";
```

This is weird. I should refactor this.

drupol replied:
Do you need me to fixup those commits?

roberth commented on flake.nix:
```diff
@@ -374,15 +388,42 @@
 
-            Generates a configuration for [pre-commit](https://pre-commit.com),
-            provides a script to activate it, and adds a [check](flake-parts.html#opt-perSystem.checks).
+            The structure is identical to the [Nixpkgs one](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md).
```
This isn't accurate; it behaves somewhat like https://nixos.org/manual/nixpkgs/stable/index.html#function-library-lib.filesystem.packagesFromDirectoryRecursive

It fulfills a similar purpose, but the details are quite different.

Just referring to aforemented function doesn't cut it, I'm afraid, because of the issue with directories _below_ `pkgsDirectory` producing an `attrsOf package` instead of a `package`. That's a type error.

roberth commented on flake.nix:
````diff
@@ -374,15 +388,42 @@
+                };
+              };
+            }
+            ```
````
I really like that this example is complete.

I'll push a commit to turn off the generated installation instructions (see [preview](https://66ed5dbe2037df4b8182b424--sprightly-cassata-33312f.netlify.app/options/pkgs-by-name-for-flake-parts))

roberth replied:
[done](https://66ed5f55322d5d4a6a131923--sprightly-cassata-33312f.netlify.app/options/pkgs-by-name-for-flake-parts)

roberth commented:
While it's not necessarily my role to thoroughly vet all contributions (they're changed externally anyway), I couldn't help but notice an apparent contradiction, and indeed when I try this, I get:

`flake.nix`
```nix
{
  inputs = {
    flake-parts.url = "github:hercules-ci/flake-parts";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    pkgs-by-name-for-flake-parts.url = "github:drupol/pkgs-by-name-for-flake-parts";
  };

  outputs = inputs@{ flake-parts, ... }:
    flake-parts.lib.mkFlake { inherit inputs; } {
      imports = [
	inputs.pkgs-by-name-for-flake-parts.flakeModule
      ];
      systems = [ "x86_64-linux" "aarch64-linux" "aarch64-darwin" "x86_64-darwin" ];
      perSystem = { config, self', inputs', pkgs, system, ... }: {
        pkgsDirectory = ./nix/pkgs-by-name;
      };
    };
}
```

shell

```console
$ (path=nix/pkgs-by-name/a/b/c/package.nix; mkdir -p $(dirname $path); echo '{ hello }: hello' >$path; git add $path)`

$ nix build .#a.b.c
[...]
       error: A definition for option `perSystem.x86_64-linux.packages.a' is not of type `package'. Definition values:
       - In `/nix/store/p872s6qyc1sv7hdx6ljl6m7rlz0l8lgm-source/flake-module.nix, via option perSystem':
           {
             b = {
               c = <derivation hello-2.12.1>;
             };
           }
```

drupol commented:
This is something that should be working indeed. If you have a pointer on how to make it work like that, please let me know!

drupol commented:
> I think you have a bunch of options here, and you could probably turn them into literal `options`.
> 
>     * Flatten the tree by calling `concatMapAttrs` recursively (or perhaps some complicated `lib` function, idk, I'd go for `concatMapAttrs`). The result would look like `.#a-b-c`, but I've also seen `.#a:b:c` or `.#a/b/c` out there.
>       * either before or after creating the scope
> 

Implemented using the original `nixpkgs`'s `lib.filesystem.packagesFromDirectoryRecursive` and a custom lazy `callPackage` callback, and a custom `flattenAttrs` function.
       
>     * Put the packages in `legacyPackages`, so that `.#a.b.c` works

Implemented using the original `nixpkgs`'s `lib.filesystem.packagesFromDirectoryRecursive`.

>     * Generate pseudo-packages `.#a` and `.#a.b` with `pkg // subpkgs`. This will make the "subpackages" unavailable after `overrideAttrs`, but that's good because making those `.b` etc packages pick up on the overriding is hard or next to impossible, esp. generically
>       
>       * does not have the word `legacy`
>       * perhaps a package like `.#a` could be useful, being a `buildEnv` or `symlinkJoin`
>       * `nix flake show` doesn't know the meaning of arbitrary attributes on individual packages, so it can't find the nested ones
>       * adds a lot of noise to the tree. You could reduce that with `lib.lazyDerivation`, removing unnecessary attrs and improving performance _slightly_; robustness a bit more.

Not implemented, I don't think at the moment this is required.

> You can tell we get a lot of trade-offs here, so this module could have a lot of `options`. I don't need you to implement all of that before merging this; just a description that's accurate or referring to "the options" when it comes to details.

My main concern is to re-use as much as possible what's already existing, this is the reason why both attributes (`legacyPackages` and `packages`) are using `lib.filesystem.packagesFromDirectoryRecursive`.

roberth commented:
Docs are looking good. I'll kick off the update after this to get the latest options for this module into the lock. It will auto update twice a day.

drupol commented:
Excellent, thanks !

roberth commented:
Thanks!
